File: frontend-preview/src/lib/api/graph.ts
/**
 * Graph API
 * Handles relation graph retrieval: ego neighborhoods, explicit subgraphs,
 * two-point paths, connected components and file impact analysis.
 *
 * All endpoints are project-scoped and return node-link structures that map
 * directly onto a graph renderer's element model. Responses carry a
 * `relation_epoch` version marker which callers should use for cache
 * invalidation. Wire types come from the generated OpenAPI contract.
 */

import { apiClient } from './client';
import type { components } from './schema';

export type GraphNode = components['schemas']['GraphNode'];
export type GraphEdge = components['schemas']['GraphEdge'];
export type GraphSubgraphResponse = components['schemas']['GraphSubgraphResponse'];
export type GraphPathResponse = components['schemas']['GraphPathResponse'];
export type GraphComponentsResponse = components['schemas']['GraphComponentsResponse'];
export type GraphImpactResponse = components['schemas']['GraphImpactResponse'];

/** Traversal direction for ego queries. */
export type GraphDirection = 'in' | 'out' | 'both';

/** Ego neighborhood query options. */
export interface EgoOptions {
	entityId: string;
	depth?: number;
	direction?: GraphDirection;
}

/** Two-point path query options. */
export interface GraphPathOptions {
	start: string;
	end: string;
	maxDepth?: number;
}

const DEFAULT_EGO_DEPTH = 2;
const DEFAULT_EGO_DIRECTION: GraphDirection = 'both';
const DEFAULT_PATH_DEPTH = 10;

/** Upper bound accepted by the backend for explicit subgraph queries. */
export const MAX_SUBGRAPH_IDS = 200;

function buildQuery(params: Record<string, string | number | undefined>): string {
	const search = new URLSearchParams();
	for (const [key, value] of Object.entries(params)) {
		if (value === undefined) continue;
		search.set(key, String(value));
	}
	const query = search.toString();
	return query ? `?${query}` : '';
}

export const graphApi = {
	/**
	 * Neighborhood around a single entity. `depth` is clamped by the backend
	 * against the project's configured relation max depth.
	 */
	getEgo: (projectId: number, options: EgoOptions) =>
		apiClient.get<GraphSubgraphResponse>(
			`/api/project/${projectId}/graph/ego${buildQuery({
				entity_id: options.entityId,
				depth: options.depth ?? DEFAULT_EGO_DEPTH,
				direction: options.direction ?? DEFAULT_EGO_DIRECTION,
			})}`
		),

	/** Shortest relation path between two entities. */
	getPath: (projectId: number, options: GraphPathOptions) =>
		apiClient.get<GraphPathResponse>(
			`/api/project/${projectId}/graph/path${buildQuery({
				start: options.start,
				end: options.end,
				max_depth: options.maxDepth ?? DEFAULT_PATH_DEPTH,
			})}`
		),

	/** Subgraph for an explicit set of stable entity ids. */
	getSubgraph: (projectId: number, ids: string[]) => {
		const limited = ids.filter((id) => id.trim().length > 0).slice(0, MAX_SUBGRAPH_IDS);
		return apiClient.get<GraphSubgraphResponse>(
			`/api/project/${projectId}/graph/subgraph${buildQuery({ ids: limited.join(',') })}`
		);
	},

	/** Connected components, used to group entities into communities. */
	getComponents: (projectId: number) =>
		apiClient.get<GraphComponentsResponse>(`/api/project/${projectId}/graph/components`),

	/**
	 * Bulk graph export. The backend defaults to a bounded limit, so callers
	 * should not assume this returns the entire project graph.
	 */
	exportGraph: (projectId: number, limit?: number) =>
		apiClient.get<GraphSubgraphResponse>(
			`/api/project/${projectId}/graph/export${buildQuery({ limit })}`
		),

	/** Dependent entities and impact score for a changed file. */
	getImpact: (projectId: number, file: string) =>
		apiClient.get<GraphImpactResponse>(
			`/api/project/${projectId}/graph/impact${buildQuery({ file })}`
		),
};

File: crates/app/cce-api/src/models/graph.rs
//! Graph query models
//!
//! Response shapes for the project-scoped graph retrieval endpoints.
//! Nodes and edges follow the node-link layout shared with the
//! orchestrator graph service.

use serde::{Deserialize, Serialize};
use utoipa::{IntoParams, ToSchema};

/// A node in a returned subgraph.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct GraphNode {
    pub id: String,
    pub label: String,
    pub kind: String,
    pub source_file: String,
    pub source_location: String,
}

/// An edge in a returned subgraph.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct GraphEdge {
    pub source: String,
    pub target: String,
    pub relation: String,
    pub confidence: String,
}

/// Ego neighborhood query parameters.
#[derive(Debug, Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub struct EgoQuery {
    pub entity_id: String,
    #[serde(default = "default_ego_depth")]
    pub depth: usize,
    #[serde(default = "default_ego_direction")]
    pub direction: String,
}

/// Two-point path query parameters.
#[derive(Debug, Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub struct GraphPathQuery {
    pub start: String,
    pub end: String,
    #[serde(default = "default_path_depth")]
    pub max_depth: usize,
}

/// Explicit entity set query parameters (comma-separated stable ids).
#[derive(Debug, Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub struct SubgraphQuery {
    pub ids: String,
}

/// Full export query parameters.
#[derive(Debug, Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub struct ExportQuery {
    #[serde(default = "default_export_limit")]
    pub limit: usize,
}

/// File impact query parameters.
#[derive(Debug, Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub struct ImpactQuery {
    pub file: String,
}

/// Subgraph response (ego, subgraph, export).
#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct GraphSubgraphResponse {
    pub success: bool,
    pub relation_epoch: i64,
    pub nodes: Vec<GraphNode>,
    pub edges: Vec<GraphEdge>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relation_info: Option<serde_json::Value>,
}

/// Path response.
#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct GraphPathResponse {
    pub success: bool,
    pub relation_epoch: i64,
    pub path_found: bool,
    #[serde(default)]
    pub nodes: Vec<GraphNode>,
    #[serde(default)]
    pub edges: Vec<GraphEdge>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relation_info: Option<serde_json::Value>,
}

/// Connected components response (stable id groups).
#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct GraphComponentsResponse {
    pub success: bool,
    pub relation_epoch: i64,
    pub components: Vec<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relation_info: Option<serde_json::Value>,
}

/// File impact response.
#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct GraphImpactResponse {
    pub success: bool,
    pub relation_epoch: i64,
    pub changed_file: String,
    pub direct_dependents: Vec<String>,
    pub transitive_dependents: Vec<String>,
    pub impact_score: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relation_info: Option<serde_json::Value>,
}

fn default_ego_depth() -> usize {
    2
}

fn default_ego_direction() -> String {
    "both".to_string()
}

fn default_path_depth() -> usize {
    10
}

fn default_export_limit() -> usize {
    2000
}
